### PR TITLE
[feat] Sha1 Store enhancements

### PR DESCRIPTION
**Pull request by min-xu-ai**
## What does this PR do?

- mainly adding ability to add not only file but also state dict and tensors to SHA1_Store.
- many small cleanup as well. Better review the individual commits so that the changes are easier to see.

Still a work in progress, will tag you Riyasat when this is ready.

## Before submitting

- [ ] Did you have fun?
  - Make sure you had fun coding 🙃
- [ ] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/main/CONTRIBUTING.md)?
- [ ] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
  - [ ] N/A
- [ ] Did you make sure to update the docs?
  - [ ] N/A
- [ ] Did you write any new necessary tests?
  - [ ] N/A
- [ ] Did you update the [changelog](https://github.com/facebookresearch/fairscale/blob/main/CHANGELOG.md)? (if needed)
  - [ ] N/A

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

### CODE REVIEW

File: fairscale/experimental/wgit/sha1_store.py [hunk 1]
```diff
@@ -4,160 +4,257 @@
...
+        # Path types, which inherit from the Path class.
+        if isinstance(file_or_obj, (Path, str)):
+            # Make sure it is a valid file.
+            torch.load(cast(Union[Path, str], file_or_obj))
+            file_path = Path(file_or_obj)
+            remove_tmp = False
+        elif isinstance(file_or_obj, (Tensor, OrderedDict)):
+            # Serialize the object into a tmp file.
+            file_path = self._get_tmp_file_path()
+            torch.save(cast(Union[Tensor, OrderedDict], file_or_obj), file_path)
```
Comment (riohib) [inline fairscale/experimental/wgit/sha1_store.py:134]: Is this casting for adhering to mypy type checking?

File: fairscale/experimental/wgit/sha1_store.py [hunk 2]
```diff
@@ -4,160 +4,257 @@
...
-    def _add_ref(self, current_sha1_hash: str, parent_hash: str) -> None:
+    def _get_tmp_file_path(self) -> Path:
+        """Helper to get a tmp file name under self.tmp_dir."""
+        return Path(tempfile.mkstemp(dir=self._tmp_dir)[1])
+
+    def _sha1_to_dir(self, sha1: str) -> Path:
+        """Helper to get the internal dir for a file based on its SHA1"""
+        # Using first 2 letters of the sha1, which results 26 * 26 = 676 subdirs under the top
+        # level. Then, using another 2 letters for sub-sub-dir. If each dir holds 1000 files, this
+        # can hold 450 millions files.
```
Comment (riohib) [inline fairscale/experimental/wgit/sha1_store.py:225]: nice to know!!

File: fairscale/experimental/wgit/sha1_store.py (hunk 1 above)
Comment (min-xu-ai) [reply fairscale/experimental/wgit/sha1_store.py:134]: Exactly😀

**Review by riohib — APPROVED** [review]
LGTM :D

File: tests/experimental/wgit/test_sha1_store.py [hunk 3]
```diff
@@ -3,148 +3,106 @@
...
+    This only runs once per test function. So don't make this too slow.
 
-@pytest.fixture
-def sha1_store(sha1_configs):
-    repo = Repo(sha1_configs.test_path.parent, init=False)
-    sha1_store = SHA1_store(sha1_configs.test_dirs, init=False)
-    return repo, sha1_store
+    Tests must be written in a way that either all of the tests run
+    in the order they appears in this file or a specific test is
+    run separately by the user. Either way, the test should work.
```
Comment (min-xu-ai) [inline tests/experimental/wgit/test_sha1_store.py:29]: this could be something to pay attention to for existing and future tests. @riohib

**Comment by min-xu-ai** [comment]
one of the unit test failure should be fixed. Another one seems like an unrelated issue.  Will see if it is flaky or not.

**Pull request merged**